File: tmb_examples/thetalog.R
library(TMB)
compile("thetalog.cpp")
dyn.load(dynlib("thetalog"))

## Read data
Y <- scan("thetalog.dat", skip=3, quiet=TRUE)
data <- list(Y=Y)

## Parameter initial guess
parameters <- list(
  X = data$Y*0,
  logr0 = 0,
  logtheta = 0,
  logK = 6,
  logQ = 0,
  logR = 0
)

## Fit model
obj <- MakeADFun(data, parameters, random="X", DLL="thetalog")
newtonOption(obj, smartsearch=FALSE)
obj$fn()
obj$gr()
system.time(opt <- nlminb(obj$par, obj$fn, obj$gr))
rep <- sdreport(obj)
rep

File: tmb_examples/thetalog.cpp
// Theta logistic population model from Pedersen et al 2012, Ecol. Modelling.
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  /* Data section */
  DATA_VECTOR(Y);
  /* Parameter section */
  PARAMETER_VECTOR(X);
  PARAMETER(logr0);
  PARAMETER(logtheta);
  PARAMETER(logK);
  PARAMETER(logQ);
  PARAMETER(logR);
  /* Procedure section */
  Type r0=exp(logr0);
  Type theta=exp(logtheta);
  Type K=exp(logK);
  Type Q=exp(logQ);
  Type R=exp(logR);
  int timeSteps=Y.size();
  Type ans=0;
  for(int i=1;i<timeSteps;i++){
    Type m=X[i-1]+r0*(1.0-pow(exp(X[i-1])/K,theta));
    ans-=dnorm(X[i],m,sqrt(Q),true);
  }
  for(int i=0;i<timeSteps;i++){
    ans-=dnorm(Y[i],X[i],sqrt(R),true);
  }
  return ans;
}
